File: Sep/Day1/maximum-average-pass-ratio.py
# TLE  ::  Time: O(K * N)  ::   Space: O(N)
class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        passRatio = []

        for class_i in classes:
            ratio = class_i[0] / class_i[1]
            passRatio.append(initial_ratio)

        while extraStudents > 0:
            updatedRatios = []

            for class_ in classes:
                ratio = (class_[0] + 1) / (class_[1] + 1)
                updatedRatios.append(ratio)

            choosenClassIndex = 0
            maxGain = 0

            for i in range(len(updatedRatios)):
                gain = updatedRatios[i] - passRatio[i]
                if gain > maxGain:
                    choosenClassIndex = i
                    maxGain = gain

            passRatio[choosenClassIndex] = updatedRatios[choosenClassIndex]
            classes[choosenClassIndex][0] += 1
            classes[choosenClassIndex][1] += 1

            extraStudents -= 1

        totalPassRatio = sum(passRatio)
        return totalPassRatio / len(classes)



# Optimized  ::  Time: O((K + N) * Log(N))  ::   Space: O(N)
class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        def gain(passedStudents, totalStudents):
            return (passedStudents + 1) / (totalStudents + 1) - passedStudents / totalStudents

        # Max heap to store (-gain, passedStudents, totalStudents)
        maxHeap = []
        for passedStudents, totalStudents in classes:
            currGain = gain(passedStudents, totalStudents)
            maxHeap.append((-currGain, passedStudents, totalStudents))

        heapq.heapify(maxHeap)
        for _ in range(extraStudents):
            currGain, passedStudents, totalStudents = heapq.heappop(maxHeap)
            heapq.heappush(
                maxHeap,
                (
                    -gain(passedStudents + 1, totalStudents + 1),
                    passedStudents + 1,
                    totalStudents + 1,
                ),
            )

        totalPassRatio = sum(
            passedStudents / totalStudents for _, passedStudents, totalStudents in maxHeap
        )
        return totalPassRatio / len(classes)

File: Sep/Day1/maximum-average-pass-ratio.cc
// TLE  ::  Time: O(K * N)  ::   Space: O(N)
class Solution {
public:
    double maxAverageRatio(vector<vector<int>>& classes, int extraStudents) {
        int n = classes.size();
        vector<double> passRatio;
        for (auto &c : classes) {
            passRatio.push_back((double)c[0] / c[1]);
        }

        while (extraStudents-- > 0) {
            vector<double> updatedRatios;
            for (auto &c : classes) {
                updatedRatios.push_back((double)(c[0] + 1) / (c[1] + 1));
            }

            int chosenIdx = 0;
            double maxGain = 0.0;
            for (int i = 0; i < n; i++) {
                double gain = updatedRatios[i] - passRatio[i];
                if (gain > maxGain) {
                    maxGain = gain;
                    chosenIdx = i;
                }
            }

            passRatio[chosenIdx] = updatedRatios[chosenIdx];
            classes[chosenIdx][0]++;
            classes[chosenIdx][1]++;
        }

        double total = 0;
        for (double r : passRatio) total += r;
        return total / n;
    }
};




// Optimized  ::  Time: O((K + N) * Log(N))  ::   Space: O(N)
class Solution {
public:
    struct ClassData {
        int pass, total;
        double gain() const {
            return (double)(pass + 1) / (total + 1) - (double)pass / total;
        }
    };

    struct Compare {
        bool operator()(const ClassData& a, const ClassData& b) {
            return a.gain() < b.gain(); // max-heap
        }
    };

    double maxAverageRatio(vector<vector<int>>& classes, int extraStudents) {
        priority_queue<ClassData, vector<ClassData>, Compare> pq;
        for (auto &c : classes) pq.push({c[0], c[1]});

        while (extraStudents-- > 0) {
            auto top = pq.top(); pq.pop();
            top.pass++;
            top.total++;
            pq.push(top);
        }

        double total = 0;
        while (!pq.empty()) {
            auto c = pq.top(); pq.pop();
            total += (double)c.pass / c.total;
        }
        return total / classes.size();
    }
};
